[tftrt]

jjsjann123 opened the pull request:
Added alignment in trt_allocator
  Since TensorFlow gpu_bfc_allocator does not abide to the alignment requested.

jjsjann123 committed:
[tftrt]
  Added alignment in trt_allocator
  Since TensorFlow gpu_bfc_allocator does not abide to the alignment requested.

jjsjann123 commented:
@aaroey This PR should resolve the virtual destructor issue, as well as the error from cudnnFusedConvActLayer.cpp Cuda Error in createFilterTextureFused: 11

rmlarsen commented on tensorflow/contrib/tensorrt/resources/trt_allocator.cc:
```diff
@@ -37,8 +37,22 @@ void TRTCudaAllocator::free(void* memory) { cudaFree(memory); }
-  void* mem = allocator_->AllocateRaw(alignment, size);
+  void* mem = allocator_->AllocateRaw(alignment, size + alignment);
+
+  CHECK(mem_pool.count(mem) == 0);
```
Instead of calling count, it would be more efficient to check the return value of insert to avoid doing two lookups, i.e.
```
CHECK(mem_pool.insert(mem).second);
```

rmlarsen commented on tensorflow/contrib/tensorrt/resources/trt_allocator.cc:
```diff
@@ -37,8 +37,22 @@ void TRTCudaAllocator::free(void* memory) { cudaFree(memory); }
+  std::align(alignment, size, mem, total_size);
+  CHECK(mem);
+  if (mem != alloc_mem) {
+    CHECK(mem_map.count(mem) == 0);
```
Ditto:
```
CHECK(mem_map.insert(mem).second);
```

rmlarsen commented on tensorflow/contrib/tensorrt/resources/trt_allocator.cc:
```diff
@@ -51,6 +65,14 @@ TRTDeviceAllocator::TRTDeviceAllocator(tensorflow::Allocator* allocator)
 void TRTDeviceAllocator::free(void* memory) {
   VLOG(2) << "Deallocating @ " << memory;
+  // allocated memory adjusted for alignment, restore the original pointer
+  if (mem_map.count(memory) != 0) {
```
It would be more efficient to implement this as:
```
auto alloc_mem_it = mem_map.find(memory);
if (alloc_mem_it != mem_map.end()) {
  memory = *alloc_mem_it;
  mem_map.erase(alloc_mem_it);   // O(1) complexity.
}
CHECK(mem_pool.erase(memory) != 0);
```

aaroey commented on tensorflow/contrib/tensorrt/resources/trt_allocator.h:
```diff
@@ -59,6 +66,10 @@ class TRTDeviceAllocator : public nvinfer1::IGpuAllocator {
   tensorflow::Allocator* allocator_;
+
+  // supporting alignment from allocation request requires a map to free;
+  std::unordered_map<void*, void*> mem_map;
```
Please add trailing underscore to the name. Same for `mem_pool`.

jjsjann123 replied:
Oops. :)

aaroey commented on tensorflow/contrib/tensorrt/convert/convert_graph.cc:
```diff
@@ -829,7 +829,7 @@ tensorflow::Status ConvertAfterShapes(ConversionParams& params) {
     // will be destroyed after we get the serialized engine string, so it's fine
     // to use unique_ptr here.
-    std::unique_ptr<nvinfer1::IGpuAllocator> alloc;
+    std::unique_ptr<TRTDeviceAllocator> alloc;
```
Can we use TRTBaseAllocator here?

jjsjann123 replied:
We could. The following code is only using TRTDeviceAllocator tho.
Will update to TRTBaseAllocator

aaroey commented on tensorflow/contrib/tensorrt/resources/trt_allocator.cc:
```diff
@@ -37,8 +37,22 @@ void TRTCudaAllocator::free(void* memory) { cudaFree(memory); }
+  CHECK(mem);
+  void* alloc_mem = mem;
+  uint64_t total_size = size + alignment;
+  std::align(alignment, size, mem, total_size);
```
Do we need to do `mem = std::align(blablabla)`? According to https://en.cppreference.com/w/cpp/memory/align, the returned value is `The adjusted value of ptr, or null pointer value if the space provided is too small`. So I think we need to check the returned result?

aaroey commented on tensorflow/contrib/tensorrt/resources/trt_allocator.cc:
```diff
@@ -51,6 +65,14 @@ TRTDeviceAllocator::TRTDeviceAllocator(tensorflow::Allocator* allocator)
+    mem_map.erase(memory);
+    memory = alloc_mem;
+  }
+  CHECK(mem_pool.count(memory) != 0);
```
Is `mem_pool` added just for validation purpose? If yes, how about using `mem_map` to track mappings from aligned address to allocated address regardless whether they're the same (currently `mem_map` only tracks address pairs whose key and value are different)?

aaroey commented on tensorflow/contrib/tensorrt/resources/trt_allocator.cc:
```diff
@@ -37,8 +37,22 @@ void TRTCudaAllocator::free(void* memory) { cudaFree(memory); }
 void* TRTDeviceAllocator::allocate(uint64_t size, uint64_t alignment,
                                    uint32_t flags) {
+  // WAR for allocator alignment requirement
+  alignment = 512;
```
Why do we need to set it to 512?

jjsjann123 replied:
Temporary WAR for TRT allocator. Some cuda API calls require allocated gpu memory to abide to cudaDeviceProp::textureAlignment. The specific call that's causing the error: https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TEXTURE__OBJECT.html

Advised by the TRT team to just override it to 512.

aaroey replied:
Thanks, would you help to add this as comments and reference #20856?

jjsjann123 committed:
addressing review comments; changing set to unordered_set

jjsjann123 committed:
1. rename class member variables with trailing underscore
2. change unique_ptr allocator type to base class.

aaroey commented on tensorflow/contrib/tensorrt/resources/trt_allocator.cc:
```diff
@@ -37,8 +37,20 @@ void TRTCudaAllocator::free(void* memory) { cudaFree(memory); }
+  void* mem = allocator_->AllocateRaw(alignment, size + alignment);
+  CHECK(mem);
+
+  CHECK(mem_pool_.insert(mem).second);
```
Is `mem_pool_` added just for validation purpose? If yes, how about using `mem_map_` to track mappings from aligned address to allocated address regardless whether they're the same (currently `mem_map_` only tracks address pairs whose key and value are different)?

jjsjann123 replied:
Yes, mem_pool_ is only for validation purpose. (maybe not necessary?)
Having an unordered_set aside from unordered_map is trying to save space.

We could: 1) remove the mem_pool_ and validation completely; or 2) if memory consumption is not of concern, we can remove the mem_pool_ and offload everything to mem_map_

aaroey commented on tensorflow/contrib/tensorrt/resources/trt_allocator.cc:
```diff
@@ -37,8 +37,20 @@ void TRTCudaAllocator::free(void* memory) { cudaFree(memory); }
+  CHECK(mem_pool_.insert(mem).second);
+  void* alloc_mem = mem;
+  uint64_t total_size = size + alignment;
+  std::align(alignment, size, mem, total_size);
```
Do we need to do `mem = std::align(blablabla)`? According to https://en.cppreference.com/w/cpp/memory/align, the returned value is `The adjusted value of ptr, or null pointer value if the space provided is too small`. So I think we need to check the returned result?

jjsjann123 replied:
Good catch, I missed this line: "The function modifies the pointer only if it would be possible to fit the wanted number of bytes aligned by the given alignment into the buffer. If the buffer is too small, the function does nothing and returns nullptr."
Would update the later CHECK(mem) into checking the return of this function instead.

jjsjann123 committed:
check the status for std::align; removing memory allocate/free validation

jjsjann123 committed:
Merge branch 'master' into alignment_pr

aaroey commented on tensorflow/contrib/tensorrt/resources/trt_allocator.cc:
```diff
@@ -37,8 +37,20 @@ void TRTCudaAllocator::free(void* memory) { cudaFree(memory); }
   assert((alignment & (alignment - 1)) == 0);  // zero or a power of 2.
-  void* mem = allocator_->AllocateRaw(alignment, size);
+  void* mem = allocator_->AllocateRaw(alignment, size + alignment);
+  CHECK(mem);
```
Is it fine to leave nullptr for trt library to handle (and return error properly)?

jjsjann123 replied:
It should be fine according to their API: https://docs.nvidia.com/deeplearning/sdk/tensorrt-api/c_api/classnvinfer1_1_1_i_gpu_allocator.html
Will remove all CHECK and return nullptr instead.

jjsjann123 committed:
[trt allocator]
  return nullptr when allocation fails
  abort free() when receiving nullptr as input

zhenyu commented:
When will be this  fix be merged to formal release than mainline?
I encounter this issue with the following env:
Build env:
Host: p3.16xlarge
Gcc: 4.8.5 (
Which gcc: /usr/bin/gcc
ll /usr/bin/gcc-> /usr/bin/gcc48
)
AMI: Deep Learning AMI (Amazon Linux) Version 12.0 (ami-45655f20)
TensorRT: tar file(4.0.16 for ubuntu 14 ) installed at /home/ec2-user/tensorrt
Tensorflow 1.10
Build config:
Cuda 9, cudnn 7.1 nvcc 1.3, jemalloc enabled , others no
 
Running env:
Host: p3.2xlarge
AMI: Deep Learning AMI (Amazon Linux) Version 12.0 (ami-45655f20)
TensorRT: tar file(4.0.16 for ubuntu 14 ) installed at /home/ec2-user/tensorrt
Tensorflow: wheel form build env
